rust/src/operations/delete.rs

```diff
@@ -0,0 +1,1074 @@
+    use arrow::record_batch::RecordBatch;
+    use datafusion::assert_batches_sorted_eq;
+    use datafusion::from_slice::FromSlice;
+    use datafusion::prelude::*;
+    use std::sync::Arc;
+
+    async fn setup_table(partitions: Option<Vec<&str>>) -> DeltaTable {
+        let table_schema = get_delta_schema();
+
+        let table = DeltaOps::new_in_memory()
+            .create()
+            .with_columns(table_schema.get_fields().clone())
+            .with_partition_columns(partitions.unwrap_or_default())
+            .await
+            .unwrap();
+        assert_eq!(table.version(), 0);
+        table
+    }
+
+    async fn get_data(table: DeltaTable) -> Vec<RecordBatch> {
+        let ctx = SessionContext::new();
+        ctx.register_table("test", Arc::new(table)).unwrap();
+        ctx.sql("select * from test")
+            .await
+            .unwrap()
+            .collect()
+            .await
+            .unwrap()
+    }
+
+    #[tokio::test]
+    async fn test_delete_default() {
+        let schema = get_arrow_schema(&None);
+        let table = setup_table(None).await;
+
+        let batch = RecordBatch::try_new(
+            Arc::clone(&schema),
+            vec![
+                Arc::new(arrow::array::StringArray::from_slice(["A", "B", "A", "A"])),
+                Arc::new(arrow::array::Int32Array::from_slice([1, 10, 10, 100])),
+                Arc::new(arrow::array::StringArray::from_slice([
+                    "2021-02-02",
+                    "2021-02-02",
+                    "2021-02-02",
+                    "2021-02-02",
+                ])),
+            ],
+        )
+        .unwrap();
+        // write some data
+        let table = DeltaOps(table)
+            .write(vec![batch.clone()])
+            .with_save_mode(SaveMode::Append)
+            .await
+            .unwrap();
+        assert_eq!(table.version(), 1);
+        assert_eq!(table.get_file_uris().count(), 1);
+
+        let (table, metrics) = DeltaOps(table).delete().await.unwrap();
+
+        assert_eq!(table.version(), 2);
+        assert_eq!(table.get_file_uris().count(), 0);
+        assert_eq!(metrics.num_added_files, 0);
+        assert_eq!(metrics.num_removed_files, 1);
+        assert_eq!(metrics.num_deleted_rows, None);
+        assert_eq!(metrics.num_copied_rows, None);
+
+        // Scan and rewrite is not required
+        assert_eq!(metrics.scan_time_ms, 0);
+        assert_eq!(metrics.rewrite_time_ms, 0);
+
+        // Deletes with no changes to state must not commit
+        let (table, metrics) = DeltaOps(table).delete().await.unwrap();
+        assert_eq!(table.version(), 2);
+        assert_eq!(metrics.num_added_files, 0);
+        assert_eq!(metrics.num_removed_files, 0);
+        assert_eq!(metrics.num_deleted_rows, None);
+        assert_eq!(metrics.num_copied_rows, None);
+    }
+
+    #[tokio::test]
+    async fn test_delete_on_nonpartiton_column() {
+        // Delete based on a nonpartition column
```

Comment:
How are null values handled? For example, if I have a column `x: [1, 2, null, 4]`, does `DELETE FROM table_name WHERE x > 2` delete just the last row, or also the third row? Seems like it's worth a unit test to verify.

Reply:
Yeah great call out.
The previous commit would delete the row. I've added a new test and changed the behavior to not delete the third record in this case. 

A record should only be deleted if the predicate evaluates to true other it is kept. null > 2 evaluates to UNKNOWN.
I've checked the spark implementation and the behavior aligns